File: python/sparkArgsParser.py
#!/usr/local/bin python

################################################################################
### Parsing the arguments passed to a spark script
##
##
##
##
##
## Simo Goshev
## Oct 30, 2019


## Arguments passed to the script:
## "$MY_SPARK_DATASET"
## "$MDB_MASTER_NODE"
## "$MDB_NODES"
## "$MDB_USER_CREDENTIALS"
## "$MDB_BACKEND_CREDENTIALS"
## "$MDB_DATABASE_NAME"

import sys
from collections import defaultdict

def extractInfo(myFile):
    ''' Read info from a file '''
    
    with open(myFile, 'r') as f:
        myInfo = f.readlines()
        myInfo = [item.replace('\n','') for item in myInfo]
    return(myInfo)


def parseArguments(myArgs):
    ''' Parsing the arguments passed to the script '''

    myInNames = ["myDataSetFile", "myDbNodePortFile", "myDbNodesFile",
                  "myUserCredFile", "myBECredFile", "myDbNameFile"]

    ## Store all file info into a dictionary
    fileDict = dict(zip(myInNames, myArgs[1::]))

    ## Retrieve info
    outDict = defaultdict(list)
    outDict['dataSet'] = fileDict['myDataSetFile']
    outDict['dbName'] = extractInfo(fileDict['myDbNameFile'])[0]
    outDict['dbNode'] = extractInfo(fileDict['myDbNodePortFile'])[0] \
      .split(':')[0]
    outDict['dbPort'] = extractInfo(fileDict['myDbNodePortFile'])[0] \
      .split(":")[1]
    outDict['dbUrl'] = ("jdbc:mysql://" + outDict['dbNode'] +
                        ":" + outDict['dbPort'] + "/" +
                        outDict['dbName'])
    outDict['dbUser'] = extractInfo(fileDict['myUserCredFile'])[0]
    outDict['dbPass'] = extractInfo(fileDict['myUserCredFile'])[1]

    outDict['dbNodes'] = extractInfo(fileDict['myDbNodesFile'])

    if len(outDict['dbNodes']) > 1:
        outDict['dbBENodes'] = outDict['dbNodes'][1::]
        outDict['dbBEUser'] = extractInfo(fileDict['myBECredFile'])[0]
        outDict['dbBEPass'] = extractInfo(fileDict['myBECredFile'])[1]
        
    return(outDict)

    

File: R/sparkArgsParser.R
#!/usr/local/bin R

################################################################################
### Parsing the arguments passed to a spark script
##
##
##
##
##
## Simo Goshev
## Oct 17, 2019


## Arguments passed to the script:
## "$MY_SPARK_DATASET"
## "$MDB_MASTER_NODE"
## "$MDB_NODES"
## "$MDB_USER_CREDENTIALS"
## "$MDB_BACKEND_CREDENTIALS"
## "$MDB_DATABASE_NAME"


parseArguments <- function(myArgs) {

    ## Name the arguments
    myArgNames <- c("myDataSetFile", "myDbNodePortFile", "myDbNodesFile",
                    "myUserCredFile", "myBECredFile", "myDbNameFile")
    
    myArgs <- as.list(myArgs)
    names(myArgs) <- myArgNames

    userConfig <- list()

    ## Extracting info from arguments
    if (myArgs$myDataSetFile != "") {
        
        userConfig <- c(userConfig, list(dataSet = myArgs$myDataSetFile))

    } else {

        stop("Location of data set must be provided", call.=FALSE)
    }


    ## Check whether will be using a database
    if (myArgs$myDbNodePortFile != "" ) {

        ## Read the dbName
        userConfig <- c(userConfig, list(dbName = readLines(myArgs$myDbNameFile)))    

        ## Define db node, port and name
        myDbNodePort <- str_split(readLines(myArgs$myDbNodePortFile), ":", simplify=TRUE)
        jdbcUrl <- paste0("jdbc:mysql://", myDbNodePort[1], ":", myDbNodePort[2], "/", userConfig$dbName)

        ## Retrieve user name and password for DB
        myUserCred <- readLines(myArgs$myUserCredFile)

        userConfig <- c(userConfig, list(dbNode = myDbNodePort[1],
                                         dbPort = myDbNodePort[2],
                                         dbUser = myUserCred[1],
                                         dbPass = myUserCred[2],
                                         dbUrl  = jdbcUrl))
        
        if (myArgs$myDbNodesFile != "") {
            
            ## Retireve back-end user name and password for DB
            myBEUserCred <- readLines(myArgs$myBECredFile)
            
            ## Add all nodes and backend credentials to the list
            userConfig <- c(userConfig, list(dbNodes = readLines(myArgs$myDbNodesFile),
                                             dbBEUser = myBEUserCred[1],
                                             dbBEPass = myBEUserCred[2]))
            
                   userConfig <- c(userConfig, list(dbBENodes = userConfig$dbNodes[-1]))
        }
    }
    return(userConfig)
}
